mod-source-record-storage-server/src/main/java/org/folio/dao/RecordDaoImpl.java

```diff
@@ -406,11 +419,95 @@ public Flowable<Record> streamRecords(Condition condition, RecordType recordType
+      }
+      if (expression instanceof AndExpression) {
+        combinedExpression.append(" and ");
+      } else if (expression instanceof OrExpression) {
+        combinedExpression.append(OR);
+      }
+      if (containsParenthesis(rightExpression)) {
+        combinedExpression.append(buildCteDistinctCountCondition(rightExpression));
+        combinedExpression.append(")");
+      }
+    }
+    return combinedExpression.toString();
+  }
+
+  private static void parseExpression(Expression expr, List<Expression> expressions) {
+    if (expr instanceof BinaryExpression binExpr) {
+      parseExpression(binExpr.getLeftExpression(), expressions);
+      parseExpression(binExpr.getRightExpression(), expressions);
+    } else if (expr instanceof Parenthesis parenthesis) {
+      if (containsParenthesis(parenthesis.getExpression())) parseExpression(parenthesis.getExpression(), expressions);
+      else expressions.add(parenthesis);
+    }
+  }
+
+  private static boolean containsParenthesis(Expression expr) {
+    if (expr instanceof Parenthesis) {
+      return true;
+    } else if (expr instanceof BinaryExpression binExpr) {
+      return containsParenthesis(binExpr.getLeftExpression()) || containsParenthesis(binExpr.getRightExpression());
+    } else {
+      return false;
+    }
+  }
+
+  private String buildCteWhereCondition(String whereExpression) throws JSQLParserException {
+    List<Expression> expressions = new ArrayList<>();
+    StringBuilder cteWhereCondition = new StringBuilder();
+
+    Expression expr = CCJSqlParserUtil.parseCondExpression(whereExpression);
+    parseExpression(expr, expressions);
+    int i = 1;
+    for (Expression expression : expressions) {
+      cteWhereCondition.append(expression.toString());
+      if (i < expressions.size()) cteWhereCondition.append(OR);
+      i++;
+    }
+    return cteWhereCondition.toString();
+  }
+
   @Override
-  public Flowable<Row> streamMarcRecordIds(ParseLeaderResult parseLeaderResult, ParseFieldsResult parseFieldsResult, RecordSearchParameters searchParameters, String tenantId) {
+  public Flowable<Row> streamMarcRecordIds(ParseLeaderResult parseLeaderResult, ParseFieldsResult parseFieldsResult,
+                                           RecordSearchParameters searchParameters, String tenantId) throws JSQLParserException {
     /* Building a search query */
+    //TODO: adjust bracets in condtion statements
+    CommonTableExpression commonTableExpression = null;
+    if (parseFieldsResult.isEnabled()) {
+      String cteWhereExpression = buildCteWhereCondition(parseFieldsResult.getWhereExpression());
+
+      Expression expr = CCJSqlParserUtil.parseCondExpression(parseFieldsResult.getWhereExpression());
+      String cteHavingExpression = buildCteDistinctCountCondition(expr);
+
+      commonTableExpression = DSL.name(CTE).as(
+        select(
+          field(MARC_ID))
+          .from(MARC_INDEXERS).join(RECORDS_LB.getName())
```

Comment:
propose to return back the join on the marc_records_tracking table when parseFieldsResult exists, to make a search on marc_indexers records with the latest version, as it was before

Reply:
completed. thanks!

Reply:
thanks!